### PR TITLE
Compare load file

### LINKED ISSUE #551: Page components should not always dispatch an action on mount
From: https://github.com/mozilla/addons-code-manager/pull/550

---

### Describe the problem and steps to reproduce it:

1. Load http://localhost:3000/en-US/browse/494431/versions/1532144
2. Click on the root folder to expand it
3. Change line 98 so that it says LOL instead of a line number (or any edit like this).

## What happened?

The component updates but the tree collapses. It looks like it's replaying all Redux actions starting with initialState.

### What did you expect to happen?

The `CodeView` component updates but the rest of the UI does not change, i.e., the tree is still expanded.

### Anything else we should know?

This now occurs because of HMR https://github.com/mozilla/addons-code-manager/pull/550. We should prevent our main components to always dispatch actions because HMR re-mounts them and we did not handle this edge case.

In addons-frontend, we handled this edge case because of (or thanks to) SSR.

### LINKED ISSUE #590: Link to a URL for viewing a single file (compare page)
This is similar to #286 but for the compare page, e.g,  https://code.addons-dev.allizom.org/en-US/compare/502955/versions/1541794...1541798/.

### PR DESCRIPTION
**Pull request by willdurand**
Fixes #590
Fixes #551
~~⚠️ depends on #593~~

---

I found new bugs after having fixed other unrelated bugs and rebased 😞 Now, both `Browse` and `Compare` components have a similar behavior and the keyboard navigation works for both as intended and without any side effect (hopefully).

It seemed easier to fix #551 as part of this patch, but it resulted in a bigger diff than expected. Sorry about that..

### CODE REVIEW

**Comment by kumar303** [comment]
> It seemed easier to fix #551 as part of this patch, but it resulted in a bigger diff than expected. Sorry about that..

It doesn't look too big. +/- 400 changes is roughly the limit, I'd say. This patch is under that.

**Comment by willdurand** [comment]
> It doesn't look too big. +/- 400 changes is roughly the limit, I'd say. This patch is under that.

Sure but the diff mixes changes for two different issues, which does not make it easy to follow. Especially the test file has changed quite a lot.

File: src/pages/Browse/index.spec.tsx [hunk 1]
```diff
@@ -480,8 +480,11 @@ describe(__filename, () => {
     expect(dispatch).not.toHaveBeenCalled();
   });
 
-  it('dispatches viewVersionFile on update if the query string contains a `path` that is different than `version.selectedPath`', () => {
-    const { renderAndUpdate, version } = setUpVersionFileUpdate({
+  // This could happen when a keyboard navigation updates the selected path. We
+  // do not want to update the selected path again, so we apply this logic to
+  // the first render (mount) only.
```
Comment (kumar303) [inline src/pages/Browse/index.spec.tsx]: Oh, you do have a comment 😄 Can you move this to the `loadData()` method? That's a better place for it.

File: src/pages/Compare/index.tsx [hunk 2]
```diff
@@ -66,45 +73,55 @@ export class CompareBase extends React.Component<Props> {
...
-      (prevProps.path && path !== prevProps.path) ||
+    if (!prevProps) {
+      if (path && path !== version.selectedPath) {
+        // We preserve the hash in the URL (if any) when we load the file from
+        // an URL that has likely be shared.
+        this.viewVersionFile(path, { preserveHash: true });
+      }
+    } else if (
+      (prevProps.version &&
+        version.selectedPath !== prevProps.version.selectedPath) ||
```
Comment (kumar303) [inline src/pages/Compare/index.tsx]: I guess nothing here is broken but we should update this someday to not use `prevProps` values for comparison. It is fragile. The best way to do it is to compare Redux mapped properties.

File: src/pages/Compare/index.spec.tsx [hunk 3]
```diff
@@ -435,63 +445,148 @@ describe(__filename, () => {
...
+      addonId,
+      baseVersionId,
+      headVersionId,
+      path,
+    });
+  });
+
+  // This could happen when a keyboard navigation updates the selected path. We
+  // do not want to update the selected path again, so we apply this logic to
+  // the first render (mount) only.
```
Comment (kumar303) [inline src/pages/Compare/index.spec.tsx]: This comment is helpful but I think it should move to `loadData()` where the logic happens.

**Comment by willdurand** [comment]
> It's a shame there is so much duplication between this and the `Browse` component.

Yes 😞

**Pull request merged**